// giveth-bridge/lib/utils.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.sendEmail = undefined;

var _nodemailer = require('nodemailer');

var _nodemailer2 = _interopRequireDefault(_nodemailer);

var _nodemailerMailgunTransport = require('nodemailer-mailgun-transport');

var _nodemailerMailgunTransport2 = _interopRequireDefault(_nodemailerMailgunTransport);

var _winston = require('winston');

var _winston2 = _interopRequireDefault(_winston);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var sendEmail = exports.sendEmail = function sendEmail(config, msg) {
    var mailApiKey = config.mailApiKey,
        mailDomain = config.mailDomain,
        mailFrom = config.mailFrom,
        mailTo = config.mailTo;


    if (!mailApiKey || !mailDomain || !mailFrom || !mailTo) {
        _winston2.default.info('not sending mail msg:', msg);
        return;
    }
    // Use Smtp Protocol to send Email
    var smtpTransport = _nodemailer2.default.createTransport((0, _nodemailerMailgunTransport2.default)({
        auth: {
            api_key: mailApiKey,
            domain: mailDomain
        }
    }));

    var mail = {
        from: mailFrom,
        to: mailTo,
        subject: 'Giveth bridge Error',
        text: msg
    };

    smtpTransport.sendMail(mail, function (error, response) {
        if (error) {
            console.log(error);
        }

        smtpTransport.close();
    });
}; /* istanbul ignore file */
